### ISSUE TITLE
Help wanted: compiling mockito with native-image

### ISSUE THREAD

**Issue by eiswind** [mockito/mockito#2435]
I am trying to compile mockito with mock-maker-proxy with graalvm native image, and I got stock with the plugin mechanism (At least the stacktrace point to that). (In preparation for having some mockito suppport in spring-native https://github.com/spring-projects-experimental/spring-native/issues/1063)

I know this maybe a little of topic, but having at least "some" mockito when running native tests made me get started with this.

What I've got so far is
- some proxy config
- some reflection config
- basic compiler args

When I compile this example
https://github.com/eiswind/native-mock-example

I get a NPE at:

```
     Caused by: java.lang.NullPointerException
       org.mockito.internal.configuration.plugins.Plugins.getStackTraceCleanerProvider(Plugins.java:26)
       org.mockito.internal.exceptions.stacktrace.StackTraceFilter.<clinit>(StackTraceFilter.java:20)
       com.oracle.svm.core.classinitialization.ClassInitializationInfo.invokeClassInitializer(ClassInitializationInfo.java:375)
       com.oracle.svm.core.classinitialization.ClassInitializationInfo.initialize(ClassInitializationInfo.java:295)
```

from which I would guess that registry itself is null, but that makes only little sense.
I added some --initialize-at-runtime for the plugins package but that didn't do it.

So for know I can only hope that this is something of interest and I really would appreciate some pointers what to try.

**Comment by TimvdLippe** [comment]
@raphw Does https://github.com/oracle/graal/issues/3969 resolve this issue as well?

**Comment by raphw** [comment]
Possibly, but this would require a detection run prior. There's not much of a point in my opinion, to run one of tests with Graal native image, though. The efficiency comes from repeated startup. That's not really our niche.

**Comment by eiswind** [comment]
I really think there is a point in running the tests with native image, as we can detect many configuration errors that happen only at runtime.

**Comment by raphw** [comment]
You are right, that's a use case. We will only be able to support Graal native image with the subclass mock maker for now, unfortunately, which requires a bunch of *add-opens* on the command line to work with Java 17. But in theory, it is definitely possible that this would work then.

**Comment by mhalbritter** [comment]
I've added support for Mockito to the GraalVM reachability metadata and got the `ProxyMockMaker` to work, [see here for more details](https://github.com/spring-projects/spring-boot/issues/32195#issuecomment-1337552932).

What is the status of the other mock makers (particularly one which allows mocking of classes)?

**Comment by raphw** [comment]
It works already: https://gist.github.com/raphw/e0f6864ddcad69630d839c2eebaad818

But it requires some prework. As far as I know, the Graal team is working on something to ease this.

**Comment by klopfdreh** [comment]
I grab this ticket up as we are currently also shifting to native images and the title of this ticket seems to be a general collector to address native image builds with Mockito.

There is a hint mentioning:

```plain
Mockito is currently self-attaching to enable the inline-mock-maker. This will no longer work in future releases of the JDK. Please add Mockito as an agent to your build as described in Mockito's documentation: https://javadoc.io/doc/org.mockito/mockito-core/latest/org.mockito/org/mockito/Mockito.html#0.3
```

The documentation says that we have to use a javaagent which is not possible with native-images. I know from opentelemetry that they created a starter project for this https://opentelemetry.io/blog/2023/spring-native/ - maybe it is possible to also create a submodule for that purpose and also to setup all the prework mentioned here: https://github.com/mockito/mockito/issues/2435#issuecomment-1340696857 so that not every user of Mockito needs to do so.

**Comment by raphw** [comment]
We could add native-image support in two ways:

1. Byte Buddy already supports the tracing agent: https://www.graalvm.org/22.0/reference-manual/native-image/Agent/ - We could extend the support for Mockito. This would require running the tests before.
2. We would need to transform the byte code of classes that should be mocked during build. This way the classes would be ready to mock and would no longer require instrumentation during runtime. This would however require a preselection of classes. This would neither allow to instrument system classes or even library classes, unless Graal offers a way to transform classes during build by now? If this is a feasible option, I prepared a branch: https://www.graalvm.org/22.0/reference-manual/native-image/Agent/

How would you want to run Mockito within a native image? Could you provide an example of a common work-flow, so that we can look into adding support for this?

**Comment by klopfdreh** [comment]
The documentation of Spring Boot can be found here: https://docs.spring.io/spring-boot/how-to/native-image/testing-native-applications.html

I think for normal java tests we need to add Mockito to the maven-surefire-plugin as mentioned in the log output, but I am not sure on how to do so on `mvn -PnativeTest test`

**Comment by raphw** [comment]
Well, this would be a Spring integration, if so. I think we should attempt to offer support for this as much as we can, but in the end Spring would need to integrate. What we can likely do is to adjust the mock maker to allow running in two stages where someone can apply the class file transformation that is needed. At runtime, we can then activate these mocks, but someone would need to run this API.

**Comment by raphw** [comment]
If anybody wants to have a look at this:

The branch https://github.com/mockito/mockito/tree/factor-out-class-file-transformer contains some work.

1. When creating the native image, all classes that should be mockable would need to run through a transformation via `InlineClassFileTransformer#transform(String,byte[])` which returns the transformed byte code.
2. At runtime, `-Dorg.mockito.aot=true` needs to be set. This can be optional in the future, but requires some more work and proper testing, but for now this will enable someone develop a prototype.

As soon as someone wants to dive into creating a prototype, I am happy to help on the Mockito side.
